src/mlpack/methods/ann/layer/layer_types.hpp

```diff
@@ -0,0 +1,117 @@
+#include <mlpack/methods/ann/convolution_rules/naive_convolution.hpp>
+#include <mlpack/methods/ann/convolution_rules/fft_convolution.hpp>
+
+namespace mlpack {
+namespace ann {
+
+template<typename InputDataType, typename OutputDataType> class AddMerge;
+template<typename InputDataType, typename OutputDataType> class Concat;
+template<typename InputDataType, typename OutputDataType> class DropConnect;
+template<typename InputDataType, typename OutputDataType> class Glimpse;
+template<typename InputDataType, typename OutputDataType> class Linear;
+template<typename InputDataType, typename OutputDataType> class LinearNoBias;
+template<typename InputDataType, typename OutputDataType> class LSTM;
+template<typename InputDataType, typename OutputDataType> class Recurrent;
+template<typename InputDataType, typename OutputDataType> class Sequential;
+template<typename InputDataType, typename OutputDataType> class VRClassReward;
+
+template<
+    typename OutputLayerType,
+    typename InputDataType,
+    typename OutputDataType
+>
+class ConcatPerformance;
+
+template<
+    typename ForwardConvolutionRule,
+    typename BackwardConvolutionRule,
+    typename GradientConvolutionRule,
+    typename InputDataType,
+    typename OutputDataType
+>
+class Convolution;
+
+template<
+    typename InputDataType,
+    typename OutputDataType
+>
+class RecurrentAttention;
+
+using LayerTypes = boost::variant<
+    Add<arma::mat, arma::mat>*,
+    AddMerge<arma::mat, arma::mat>*,
+    BaseLayer<LogisticFunction, arma::mat, arma::mat>*,
+    BaseLayer<IdentityFunction, arma::mat, arma::mat>*,
+    BaseLayer<TanhFunction, arma::mat, arma::mat>*,
+    BaseLayer<RectifierFunction, arma::mat, arma::mat>*,
+    Concat<arma::mat, arma::mat>*,
+    ConcatPerformance<NegativeLogLikelihood<arma::mat, arma::mat>,
+                      arma::mat, arma::mat>*,
+    Constant<arma::mat, arma::mat>*,
+    Convolution<NaiveConvolution<ValidConvolution>,
+                NaiveConvolution<FullConvolution>,
+                NaiveConvolution<ValidConvolution>, arma::mat, arma::mat>*,
+    DropConnect<arma::mat, arma::mat>*,
+    Dropout<arma::mat, arma::mat>*,
+    Glimpse<arma::mat, arma::mat>*,
+    HardTanH<arma::mat, arma::mat>*,
+    Join<arma::mat, arma::mat>*,
+    LeakyReLU<arma::mat, arma::mat>*,
+    Linear<arma::mat, arma::mat>*,
+    LinearNoBias<arma::mat, arma::mat>*,
+    LogSoftMax<arma::mat, arma::mat>*,
+    Lookup<arma::mat, arma::mat>*,
+    LSTM<arma::mat, arma::mat>*,
+    MaxPooling<arma::mat, arma::mat>*,
+    MeanPooling<arma::mat, arma::mat>*,
+    MeanSquaredError<arma::mat, arma::mat>*,
+    MultiplyConstant<arma::mat, arma::mat>*,
+    NegativeLogLikelihood<arma::mat, arma::mat>*,
+    Recurrent<arma::mat, arma::mat>*,
+    RecurrentAttention<arma::mat, arma::mat>*,
+    ReinforceNormal<arma::mat, arma::mat>*,
+    Select<arma::mat, arma::mat>*,
+    Sequential<arma::mat, arma::mat>*,
+    VRClassReward<arma::mat, arma::mat>*
+>;
```

Comment:
I was quite surprised, when I reviewed the visitor paradigm and then the boost::variant sources, to discover that boost::variant and static_visitor are completely implemented at compile time and there is no runtime overhead.  It took a while to read the boost::variant code to verify (and I don't completely understand it, only "sort of"), and then taking a look at mlpack programs:

```
$ nm -g bin/mlpack_knn | c++filt | grep MonoSearchVisitor
$
```

I'm quite impressed!  I didn't think it could be done entirely at compile-time.

Anyway, one drawback of this approach is that it doesn't make it easy for users to add their own layers.  One way you could do that is this:

```
using LayerTypes = boost::variant<
    Add<arma::mat, arma::mat>*,
    ...,
#ifdef EXTRA_LAYER_TYPES
    EXTRA_LAYER_TYPES,
#endif
    VRClassReward<arma::mat, arma::mat>*
>;
```

And then a user can add their own layer by adding that to `EXTRA_LAYER_TYPES` before including mlpack.

Another possible issue that I have not thought of any solution to is this: if you know the structure of your network at compile time you could specify a layer as, e.g., `Linear<arma::mat::fixed<10, 10>, arma::mat::fixed<10, 10>>`, but that's not part of the `boost::variant` object, so you couldn't use it.  Maybe in situations like that we can use dynamic dispatch of some sort?  I'll keep thinking but I don't know if I will be able to come up with something straightforward---`boost::any` wouldn't be a good solution because then we lose the static visitor and any call to any layer will have to be a lookup somehow.  Either way I think even if we don't have a solution for now, it would be great to use this as-is and try to improve later, because this whole idea is so much nicer for users than before too.

(I haven't reviewed the rest of the code yet---I've primarily focused only on this point so far.  I'll add comments about the rest of the code hopefully in the next few days.)